### ISSUE TITLE
gentype_getbuffer undeclared

### ISSUE THREAD

**Issue by Dekken** [numpy/numpy#13281]
Python3.5 

```
python3 -m pip install numpy ==1.15
```
### Error message:
```
    numpy/core/src/multiarray/scalartypes.c.src:2529:5: error: ‘gentype_getbuffer’ undeclared here (not in a function)
         gentype_getbuffer,                           /* bf_getbuffer */
         ^~~~~~~~~~~~~~~~~
    numpy/core/src/multiarray/scalartypes.c.src:2529:5: error: ‘gentype_getbuffer’ undeclared here (not in a function)
         gentype_getbuffer,                           /* bf_getbuffer */
```

**Comment by mattip** [comment]
What OS, hardware, compiler, external libraries, and architecture are you using? Also please show all of the error message.

**Comment by Dekken** [comment]
Debian 9 / i7-7700k / GCC 6.3 / ? / x64_86

[gist](https://gist.github.com/Dekken/4c58d1a4852cd9b3a41ab3c273127b86)

**Comment by charris** [comment]
Why is it trying to install a source file? Indeed, it looks like it is trying to install from a repo. Also, I think there is an extra space in `numpy ==1.15`

**Comment by Dekken** [comment]
```numpy ==1.15``` that was a typo

it's installing from pip

**Comment by charris** [comment]
It should be installing a binary wheel. Try 'numpy==1.15.4'

**Comment by Dekken** [comment]
I've built python from source with valgrind/system malloc so maybe that makes it build from source?

**Comment by Dekken** [comment]
configure something like this 

```
./configure CFLAGS="-g3 -O3 -march=native -fPIC -I/usr/include/openssl" CXXFLAGS="-g3 -O3 -march=native -fPIC -I/usr/include/openssl" --enable-shared LDFLAGS="-L/usr/lib -L/usr/lib/x86_64-linux-gnu -Wl,-rpath=/usr/lib/x86_64-linux-gnu" --prefix=$PWD --with-valgrind  --with-ensurepip=install
```

"--enable-optimizations" is disabled as one of the "test_socket" tests causes my machine to crash with OOM

**Comment by charris** [comment]
So probably an unrecognized Python version. You have cython installed? Try `1.15..4` as the version instead of `1.15`.

**Comment by Dekken** [comment]
sure will get back to you, the (gist)[https://gist.github.com/Dekken/4c58d1a4852cd9b3a41ab3c273127b86] uses 16.2 tho if it matters

**Comment by Dekken** [comment]
1.15.4 is the same as 1.16.2
[gist](https://gist.github.com/Dekken/b9a0e61cd74379e02853e3cd1d427113)

**Comment by Dekken** [comment]
my actual config

./configure CFLAGS="-g3 -O3 -march=native -fPIC -I/usr/include/openssl" CXXFLAGS="-g3 -O3 -march=native -fPIC -I/usr/include/openssl" --enable-shared LDFLAGS="-L/usr/lib -L/usr/lib/x86_64-linux-gnu -Wl,-rpath=/usr/lib/x86_64-linux-gnu" --prefix=/opt/py/python-3.5.6d --enable-shared --with-ensurepip=install --with-computed-gotos --enable-loadable-sqlite-extensions --without-pymalloc --with-valgrind

**Comment by charris** [comment]
I've looked as the pip source file for 1.16.2 and `numpy/core/src/multiarray/buffer.h` is there, that is where the functions are declared. But I would have thought there would be an import error if that file could not be found. In any case, there may be something missing from the include file path used by gcc. Maybe Debian is messing with pip? Where did you get the Python source?

**Comment by charris** [comment]
Does python3 point to your home built Python? Might try upgrading pip.

**Comment by Dekken** [comment]
```
philix@fad:/opt/py/python-3.5.6d$ ./python -m pip install pip --upgrade
Requirement already up-to-date: pip in ./lib/python3.5/site-packages (19.0.3)
```

**Comment by Dekken** [comment]
https://www.python.org/ftp/python/3.5.6/Python-3.5.6.tar.xz

**Comment by charris** [comment]
Hmm. Could you try adding the `--no-cache` flag, and if that doesn't work try with the system Python, maybe with `--user`. Two other things I'm curious about, first try `python3 -mpip download numpy==1.16.2` and check that `buffer.h` is there, and then try a simple `python3 setup.py install --user` from the unpacked package. I'm pretty sure that this problem is due to the Python version, but don't why.

**Comment by Dekken** [comment]
I can't use the system python, I need valgrind

**Comment by charris** [comment]
I'm wondering if the Python declaration of `PyBufferProcs` is missing.

**Comment by Dekken** [comment]
feel free to give me something to test out

**Comment by Dekken** [comment]
if --with-valgrind is used, I would wager pymalloc is disabled

**Comment by Dekken** [comment]
it looks to be pymalloc related

**Comment by ZelphirKaltstahl** [comment]
Getting the same kind of issue here, inside of a Debian Stretch Docker container:

```
    compile options: '-DNPY_INTERNAL_BUILD=1 -DHAVE_NPY_CONFIG_H=1 -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE=1 -D_LARGEFILE64_SOURCE=1 -DHAVE_CBLAS -Ibuild/src.linux-x86_64-3.6/numpy/core/src/umath -Ibuild/src.linux-x86_64-3.6/numpy/core/src/npymath -Ibuild/src.linux-x86_64-3.6/numpy/core/src/common -Inumpy/core/include -Ibuild/src.linux-x86_64-3.6/numpy/core/include/numpy -Inumpy/core/src/common -Inumpy/core/src -Inumpy/core -Inumpy/core/src/npymath -Inumpy/core/src/multiarray -Inumpy/core/src/umath -Inumpy/core/src/npysort -I/usr/local/include/python3.6m -Ibuild/src.linux-x86_64-3.6/numpy/core/src/common -Ibuild/src.linux-x86_64-3.6/numpy/core/src/npymath -Ibuild/src.linux-x86_64-3.6/numpy/core/src/common -Ibuild/src.linux-x86_64-3.6/numpy/core/src/npymath -c'
    gcc: build/src.linux-x86_64-3.6/numpy/core/src/multiarray/scalartypes.c
    numpy/core/src/multiarray/scalartypes.c.src: In function ‘gentype_dealloc’:
    numpy/core/src/multiarray/scalartypes.c.src:142:5: warning: implicit declaration of function ‘_dealloc_cached_buffer_info’ [-Wimplicit-function-declaration]
         _dealloc_cached_buffer_info(v);
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~
    numpy/core/src/multiarray/scalartypes.c.src: At top level:
    numpy/core/src/multiarray/scalartypes.c.src:2526:5: error: ‘gentype_getbuffer’ undeclared here (not in a function)
         gentype_getbuffer,                           /* bf_getbuffer */
         ^~~~~~~~~~~~~~~~~
    error: Command "gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -I/usr/local/include -I/usr/local/include/openssl -fPIC -DNPY_INTERNAL_BUILD=1 -DHAVE_NPY_CONFIG_H=1 -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE=1 -D_LARGEFILE64_SOURCE=1 -DHAVE_CBLAS -Ibuild/src.linux-x86_64-3.6/numpy/core/src/umath -Ibuild/src.linux-x86_64-3.6/numpy/core/src/npymath -Ibuild/src.linux-x86_64-3.6/numpy/core/src/common -Inumpy/core/include -Ibuild/src.linux-x86_64-3.6/numpy/core/include/numpy -Inumpy/core/src/common -Inumpy/core/src -Inumpy/core -Inumpy/core/src/npymath -Inumpy/core/src/multiarray -Inumpy/core/src/umath -Inumpy/core/src/npysort -I/usr/local/include/python3.6m -Ibuild/src.linux-x86_64-3.6/numpy/core/src/common -Ibuild/src.linux-x86_64-3.6/numpy/core/src/npymath -Ibuild/src.linux-x86_64-3.6/numpy/core/src/common -Ibuild/src.linux-x86_64-3.6/numpy/core/src/npymath -c build/src.linux-x86_64-3.6/numpy/core/src/multiarray/scalartypes.c -o build/temp.linux-x86_64-3.6/build/src.linux-x86_64-3.6/numpy/core/src/multiarray/scalartypes.o -MMD -MF build/temp.linux-x86_64-3.6/build/src.linux-x86_64-3.6/numpy/core/src/multiarray/scalartypes.o.d" failed with exit status 1
    ----------------------------------------
```

* NumPy version: `1.16.2`
* Debian Stretch: `@sha256:397b2157a9ea8d7f16c613aded70284292106e8b813fb1ed5de8a8785310a26a`

Inside a virtual machine of Xubuntu 18.04 it works without problems.

**Comment by wdirons** [comment]
Just leaving a comment here in case someone else encountered this error and did the same thing I did.

I just installed openssl from source, prior to installing python. I did export `export CPPFLAGS="-I/usr/local/include -I/usr/local/include/openssl"` and was picking up `/root/openssl-1.1.1c/include/openssl/buffer.h`, instead of the correct buffer.h that is part of numpy.

If you hit this defect, try `find / -name buffer.h` and see what other buffer.h it might be picking up.
